### DISCUSSION TITLE
improve ParamConverterProvider (bug?)

### DISCUSSION THREAD

**Discussion by redmitry** [resteasy/resteasy#3589] [General]
Hello,

I am facing a problem with multiple parameters (Wildfly 27 /RestEasy 6.2.1).
I have an endpoint (simplified here):

> GET(@QueryParam(value = "filters") List&lt;QueryFilter&gt; filters)

Which should get params like: 

> ?filters=A,B,C&filters=D

So if I implement the converter as:
`MyParamConverter implements ParamConverter<QueryFilter>`
I can't return three filters from just one parameter, so I implemented
`MyParamConverter implements ParamConverter<List<QueryFilter>>`

Here goes the problem.
RestEasy doesn't consider a possibility of collections for ParamCnvertors:
In the [StringParameterInjector.java](https://github.com/resteasy/resteasy/blob/8c5207432d34b6a6e031a1e6574475019d1eac0c/resteasy-core/src/main/java/org/jboss/resteasy/core/StringParameterInjector.java#L549) it just exit from the initialization.

Later on, in `extractValues()` it goes for a single parameter [`return extractValue(values.get(0))`](https://github.com/resteasy/resteasy/blob/8c5207432d34b6a6e031a1e6574475019d1eac0c/resteasy-core/src/main/java/org/jboss/resteasy/core/StringParameterInjector.java#L740) so I got only 'A,B,C'.

I am not sure if this is a bug, because the notation **param=A,B,C** is just one parameter in HTTP, so such queries are part of the application and not implementation library...

Kind regards,

Dmitry

**Comment by jamezp** [comment]
Do clarify, you'd expect to get "A,B,C,D"? If you've got a reproducer, which I realize might not be that simple, that would be helpful to understand if this is a bug or not.

**Reply by redmitry** [reply]
[resteasy-test.zip](https://github.com/resteasy/resteasy/files/11375631/resteasy-test.zip)

Here is the simplest example I can do.

Cheers,

D.

P.S.

Yes, I'd like to get A,B,C,D with
curl "http://localhost:8080/test?filters=A,B,C&filters=D"

**Reply by jamezp** [reply]
Sorry for the delay. This looks like a bug to me and I've filed https://issues.redhat.com/browse/RESTEASY-3330.

**Reply by jamezp** [reply]
Actually, I'm not sure this is a bug. There is a simple way around this. In your `TestFiltersConverterProvider` simply change it to:
```java
@Override
public <T> ParamConverter<T> getConverter(
        Class<T> clazz, Type genericType, Annotation[] annotations) {
    if (clazz.isAssignableFrom(TestQueryFilter.class)) {
        return (ParamConverter<T>) converter;
    }
    return null;
}
```

What will happen here is RESTEasy will see it's a collection and handle it appropriately. If this doesn't work for you let me know. We could look at improving it, but this should work.

**Comment by redmitry** [comment]
That was my initial attempt before going to the list.
>MyParamConverter implements ParamConverter&lt;QueryFilter&gt;

The problem is that I'll get two parameters (totally correct) "A,B,C" and "D" and... how would I return three params out of one?

Cheers,

Dmitry

**Reply by jamezp** [reply]
I assume your real use case is more complicated than the reproducer. In the reproducer it works because we simply just return "A,B,C,D". However, I could see in a real use case where it could get more complicated. Like returning a new instance of something for each parameter.

I'm trying to think of how we could do this.

**Comment by redmitry** [comment]
>I assume your real use case is more complicated than the reproducer.

Exactly. The filters are objects with different fields where 'A','B' are identifiers.
For now I just inject urlInfo and take all filters directly from the query. It works, but not pretty.

IMHO the solution in resteasy would be to loop over filters and collect them with addAll...

Kind regards,

Dmitry

**Reply by jamezp** [reply]
> IMHO the solution in resteasy would be to loop over filters and collect them with addAll...

But then what gets returned? The return type is an `Object` and that object needs to be readable from the readers/writers. What I think we need is for the `isCollection` to be `true`. However, since we find a `ParamConverter` the check is never done because we're told there is a `ParamConverter` that can be used. I'll have to look at how this could be done.
